File: cs214/projects/proj09/clojure/src/searchlist.clj
;; searchlist.clj tests a function to search a list for a given value.
;;
;; Output: four lists, and 
;;          the results of calling a search() function on them.
;;
;; Usage: clojure -m max
;;
;; Begun by: Dr. Adams, CS 214 at Calvin College.
;; Completed by: Advait Scaria
;; Date: 04/23/2019
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(ns searchlist)

(load "listOps")                      ; load module

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Function -main tests function search(). 
;; Output: the results of testing search() on various lists.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn -main []
  (let
    [ 
      list1 '(99 88 77 66 55)        ; 99 is first value
      list2 (list 55 66 77 88 99)    ; 99 is last value
      list3 (list 55 77 99 88 66)    ; 99 is middle value
      list4 '(5 6 7 8 9)             ; doesn't contain 99
    ]

    (println)																								; print:
    (println "list1: " list1)        ;   list1,
    (println "list2: " list2)        ;   list2,
    (println "list3: " list3)        ;   list3,
    (println "list4: " list4)    				;   list4,
    (println)                        ;   and positions of 99
    (println "Position of 99 in list1: " (search list1 99))
    (println "Position of 99 in list2: " (search list2 99))
    (println "Position of 99 in list3: " (search list3 99))
    (println "Position of 99 in list4: " (search list4 99))
    (println)
  )
)

File: cs214/projects/proj09/ruby/searchlist.rb
# searchlist.rb tests the search subprogram given a Ruby linked list.
#
# Begun by: Dr. Adams for CS 214 at Calvin College.
# Completed by: Advait Scaria
# Date: 04/23/2019

require './List'

   # search() searches a given linked list for a given value
   # Receive: aList, a list of integers; and
   #          aValue, an integer.
   # Return: the position of aValue within aList, if aValue is present in aList;
   #         -1, otherwise.

   def search(aList, aValue)
      temp = aList.instance_variable_get(:@first)
      counter = 0
      while !temp.nil?
        if temp.value == aValue
          return counter
        end
        counter = counter + 1
        temp = temp.next
      end
      return -1
   end

list1 = List.new
list2 = List.new
list3 = List.new
list4 = List.new

list1.append(99).append(88).append(77).append(66).append(55)
list2.append(55).append(66).append(77).append(88).append(99)
list3.append(55).append(77).append(99).append(88).append(66)
list4.append(5).append(6).append(7).append(8).append(9)

puts
print "list 1: "; puts list1.print
print "list 2: "; puts list2.print
print "list 3: "; puts list3.print
print "list 4: "; puts list4.print
puts
print "Position of 99 in list1: "; puts search(list1, 99)
print "Position of 99 in list2: "; puts search(list2, 99)
print "Position of 99 in list3: "; puts search(list3, 99)
print "Position of 99 in list4: "; puts search(list4, 99)
puts
